text/0000-uisize.md

```diff
@@ -0,0 +1,128 @@
+- Feature Name: `uptr_iptr`
+- Start Date: 2016-06-02
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+`isize` and `usize` either must be a `size_t` like type, or `intptr_t` like
+type. Most uses of `usize` and `isize` are as `size_t` or `ptrdiff_t`
+like-types (`Vec<T>`, `[T]`, `.offset`, etc.). Therefore, remove the guarantee
+that `usize` and `isize` are pointer-size, and make them exclusively
+object-sized.
+
+# Motivation
+[motivation]: #motivation
+
+We want to support the embedded space, and new CPUs like the CHERI. These CPUs
+do not support `usize` == `uptr`, and, in the case of the CHERI, don't support
+`uptr` at all. Most CPUs don't actually support the idea that `uptr == usize`:
+just the currently popular ones.
+
+We really want to support usecases where `uptr != usize`. The embedded space,
+where segmented architectures are still very popular, is a very good case for
```

Comment:
AVR has separate 'program' and 'data' memory, but requires separate instructions to load and store each kind, so it would be poorly suited to an implementation that sticks segment information inside pointers.

PIC18 has actual memory banks, so you could hypothetically compile code for it this way.  But I'm talking about an 8-bit architecture with a memory space of 4 bits of bank index plus 8 bits of address, for a maximum of 2^12=4096 bytes of RAM, *total*; most actual chips have significantly less.  It makes zero sense to run *cross-platform* Rust code on such a device, which is what matters for the purposes of this discussion.  (It is only barely big enough to justify the overhead of C compared to assembly, sometimes.)  Bigger PICs don't use bank switching.

These days I hear there's rarely all that much advantage (in power or cost) to using 8-bit MCUs instead of a 32-bit ARM Cortex-M0 for *any* applications, let alone ones involving programs big enough for the ability to use cross-platform code to matter.